dbms/src/TiDB/Schema/SchemaBuilder.cpp

```diff
@@ -901,29 +907,72 @@ void SchemaBuilder<Getter, NameMapper>::applyCreateSchema(const TiDB::DBInfoPtr
 }
 
 template <typename Getter, typename NameMapper>
-void SchemaBuilder<Getter, NameMapper>::applyDropSchema(DatabaseID schema_id)
+void SchemaBuilder<Getter, NameMapper>::applyRecoverDatabase(DatabaseID database_id)
 {
-    TiDB::DBInfoPtr db_info = databases.getDBInfo(schema_id);
+    auto db_info = getter.getDatabase(database_id);
     if (unlikely(db_info == nullptr))
     {
-        LOG_INFO(log, "Try to drop database but not found, may has been dropped, database_id={}", schema_id);
+        LOG_INFO(
+            log,
+            "Recover database is ignored because database is not exist in TiKV,"
+            " database_id={}",
+            database_id);
+        return;
+    }
+    LOG_INFO(log, "Recover database begin, database_id={}", database_id);
+    auto db_name = name_mapper.mapDatabaseName(database_id, keyspace_id);
+    auto db = context.tryGetDatabase(db_name);
+    if (unlikely(!db))
```

Comment:
I'm a little worried about this scenario: the customer is stuck in the gc time to do the flash back, and tiflash may receive and process it later, which may be just longer than tiflash's local gc, so the db and tables still don't exist.

Reply:
1. the tombstone_ts for `IDatabase` and `IStorage` are allocated by tiflash when it receives the `DROP DATABASE` or `DROP TABLE` DDL. So the tombstone_ts are usually sliently larger than the user actually executed those DROP statements. When a user execute `FLASHBACK DATABASE`, those `IDatabase` and `IStorage` instance are possible not pass the gc safepoint.
2. If those `IStorage` and `IDatabase` instance are actually get physically dropped in TiFlash instance, it is dangerous to "create" them when applying "FLASHBACK DATABASE". Because there are some rows is permanently deleted from tiflash instance. If we silently "create" them, it means the error is hard to find out but data inconsistency happens. An explicit error is better than implicit data loss.